core/src/test/java/org/apache/calcite/test/RexProgramTest.java

```diff
@@ -1703,6 +1703,26 @@ private void checkExponentialCnf(int n) {
     assertThat(result.getOperands().get(2), is((RexNode) falseLiteral));
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-2455">[CALCITE-2455]
+   * simplifyCoalesce of constant should match nullability</a>. */
+  @Test public void testSimplifyCoalesceWithConstant() {
+    RexNode inputRef1 = rexBuilder.makeInputRef(
+        typeFactory.createTypeWithNullability(
+            typeFactory.createSqlType(SqlTypeName.VARCHAR), true), 0);
+    RexNode inputRef2 = rexBuilder.makeInputRef(
+        typeFactory.createTypeWithNullability(
+            typeFactory.createSqlType(SqlTypeName.VARCHAR), true), 1);
+    RexCall coalesceNode = (RexCall) coalesce(inputRef2, inputRef1);
+
+    ImmutableList.Builder<RexNode> newOperands = ImmutableList.builder();
+    newOperands.add(rexBuilder.makeLiteral("S")).add(coalesceNode.operands.get(1));
+    coalesceNode = coalesceNode.clone(coalesceNode.getType(), newOperands.build());
+    RexNode result = simplify.simplify(coalesceNode);
```

Comment:
Note: there's `simplify` and `simplifyPreservingType`.
It looks like it is important to add tests for both cases so we know how each of those behave.